### PR TITLE
core: store contract IDs instead of hashes for NEP5Balances and Transfers

### LINKED ISSUE #1194: Make nep5balances store contract IDs instead of hashes
I think we can improve ae9658a10838b4616dd3206bb7d1b66818cac1e9 for Neo 3 as every contract now has a stable ID and we can store it instead of hash (getting the hash from ID when needed). This avoids migration completely.

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Closes #1194

### CODE REVIEW

File: pkg/core/dao/dao.go [hunk 1]
```diff
@@ -163,7 +164,13 @@ func (dao *Simple) GetContractState(hash util.Uint160) (*state.Contract, error)
 // PutContractState puts given contract state into the given store.
 func (dao *Simple) PutContractState(cs *state.Contract) error {
 	key := storage.AppendPrefix(storage.STContract, cs.ScriptHash().BytesBE())
-	return dao.Put(cs, key)
+	if err := dao.Put(cs, key); err != nil {
+		return err
+	}
+	key = key[:5]
```
Comment (fyrchik) [inline pkg/core/dao/dao.go:170]: I think it is better to make this a separate `dao.PutContractScriptHash` method. Hash is needed to be updated only during migration, not upon every `PutContractState`

File: pkg/rpc/server/server.go [hunk 2]
```diff
@@ -508,7 +508,11 @@ func (s *Server) getNEP5Balances(ps request.Params) (interface{}, *response.Erro
 	}
 	if as != nil {
 		cache := make(map[util.Uint160]int64)
-		for h, bal := range as.Trackers {
+		for id, bal := range as.Trackers {
+			h, errGetHash := s.chain.GetContractScriptHash(id)
```
Comment (fyrchik) [inline pkg/rpc/server/server.go:512]: Why not `err`?

File: pkg/rpc/server/server.go [hunk 3]
```diff
@@ -542,13 +542,17 @@ func (s *Server) getNEP5Transfers(ps request.Params) (interface{}, *response.Err
 	lg := s.chain.GetNEP5TransferLog(u)
 	cache := make(map[util.Uint160]int64)
 	err = lg.ForEach(func(tr *state.NEP5Transfer) error {
+		h, errGetHash := s.chain.GetContractScriptHash(tr.Asset)
```
Comment (fyrchik) [inline pkg/rpc/server/server.go:545]: `err` ?

File: pkg/rpc/server/server.go [hunk 4]
```diff
@@ -542,13 +542,17 @@ func (s *Server) getNEP5Transfers(ps request.Params) (interface{}, *response.Err
...
+		}
 		transfer := result.NEP5Transfer{
 			Timestamp: tr.Timestamp,
-			Asset:     tr.Asset,
+			Asset:     h,
 			Index:     tr.Block,
 			TxHash:    tr.Tx,
 		}
-		d, err := s.getDecimals(tr.Asset, cache)
+		d, err := s.getDecimals(h, cache)
```
Comment (fyrchik) [inline pkg/rpc/server/server.go:555]: We can impove this to make `cache` store decimals by `ID` (in a separate commit).

File: pkg/core/dao/dao.go (hunk 1 above)
Comment (AnnaShaleva) [reply pkg/core/dao/dao.go:170]: We should also *put* hash into the storage while creating/adding new contract. Each time we add new contract to the storage, the pair `hash-ID` should also be stored. We should also update this pair each time contract hash is updated. As a result, the **only** place in **the whole project** when we don't have to update `hash-ID` pair is when we [update contract manifest during migration](https://github.com/nspcc-dev/neo-go/blob/master/pkg/core/interop_neo.go#L161) (which is quite rear event).

If we make a separate `dao.PutContractScriptHash`  method, we'll have to call it almost each time after `dao.PutContractState`. So do we really want to split this methods?

I also thought about changing the name of `dao.PutContractState` to `dao.PutContractStateAndID` or smth similar.

File: pkg/rpc/server/server.go (hunk 4 above)
Comment (AnnaShaleva) [reply pkg/rpc/server/server.go:555]: I thought about that but the case is that even if we use ID for storing decimals, we'll still have to get the contract hash to fill `result.NEP5Transfer.Asset` field. So basically, there's not so mach difference between `ID` and `hash` here. Do we really need to change this?

File: pkg/core/dao/dao.go (hunk 1 above)
Comment (fyrchik) [reply pkg/core/dao/dao.go:170]: Ok, but it still can be a separate function (may be private).

File: pkg/rpc/server/server.go (hunk 4 above)
Comment (fyrchik) [reply pkg/rpc/server/server.go:555]: This will allow to cache hashes too. May be a separate issue, though.

File: pkg/rpc/server/server.go (hunk 2 above)
Comment (AnnaShaleva) [reply pkg/rpc/server/server.go]: Fixed.

File: pkg/rpc/server/server.go (hunk 3 above)
Comment (AnnaShaleva) [reply pkg/rpc/server/server.go]: Fixed.

File: pkg/rpc/server/server.go (hunk 4 above)
Comment (AnnaShaleva) [reply pkg/rpc/server/server.go]: Somehow I didn't pay attention to hashes caching. Refactored `getDecimals` in a separate commit, now it stores decimals by contract ID and caches contracts hashes also.

File: pkg/core/blockchain.go [hunk 5]
```diff
@@ -715,15 +715,26 @@ func (bc *Blockchain) processNEP5Transfer(cache *dao.Cached, h util.Uint256, b *
 		Timestamp: b.Timestamp,
 		Tx:        h,
 	}
+	var id int32
+	nativeContract := bc.contracts.ByHash(sc)
+	if nativeContract != nil {
+		id = nativeContract.Metadata().ContractID
+	} else {
+		assetContract := bc.GetContractState(sc)
```
Comment (fyrchik) [inline pkg/core/blockchain.go:715]: Do we need to have a separate case for natives?

**Pull request merged**